src/bun.js/api/bun/socket.zig

```diff
@@ -1354,35 +1354,36 @@ fn NewSocket(comptime ssl: bool) type {
             }
         }
 
-        pub fn markInactive(this: *This, socket_ctx: ?*uws.SocketContext) void {
+        pub fn markInactive(this: *This) void {
             if (this.flags.is_active) {
-                if (!this.flags.detached) {
-                    // we have to close the socket before the socket context is closed
-                    // otherwise we will get a segfault
-                    // uSockets will defer freeing the TCP socket until the next tick
-                    if (!this.socket.isClosed()) {
-                        this.socket.close(.normal);
-                        // onClose will call markInactive again
-                        return;
-                    }
+                // we have to close the socket before the socket context is closed
+                // otherwise we will get a segfault
+                // uSockets will defer freeing the TCP socket until the next tick
+                if (!this.socket.isClosed()) {
+                    this.socket.close(.normal);
+                    // onClose will call markInactive again
+                    return;
                 }
+
                 this.flags.is_active = false;
                 const vm = this.handlers.vm;
-                this.handlers.markInactive(ssl, socket_ctx, this.wrapped);
+                this.handlers.markInactive();
                 this.poll_ref.unref(vm);
                 this.has_pending_activity.store(false, .release);
             }
         }
 
         pub fn onOpen(this: *This, socket: Socket) void {
+            log("onOpen {} {}", .{ this.socket == .detached, this.ref_count });
             // update the internal socket instance to the one that was just connected
-            this.socket = socket;
+            // This socket must be replaced because the previous one is a connecting socket not a uSockets socket
+            this.socket = .{ .socket = socket };
             JSC.markBinding(@src());
             log("onOpen ssl: {}", .{comptime ssl});
 
             // Add SNI support for TLS (mongodb and others requires this)
             if (comptime ssl) {
-                var ssl_ptr = this.socket.ssl();
+                var ssl_ptr = this.socket.ssl() orelse unreachable;
```

Comment:
this is not accurate right? don't we set ->ssl = nullptr somewhere?

Reply:
If is detached